plugin/src/main/java/org/opensearch/ml/action/agents/DeleteAgentTransportAction.java

```diff
@@ -124,6 +124,7 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<Delete
                                     if (TenantAwareHelper
                                         .validateTenantResource(mlFeatureEnabledSetting, tenantId, mlAgent.getTenantId(), actionListener)) {
                                         if (mlAgent.getIsHidden() && !isSuperAdmin) {
+                                            context.restore();
```

Comment:
Or should we stash the context again before the deletion?

Reply:
This particular branch doesn't do the deletion, it's returning the 403 message because you are not superadmin and the model is hidden.  The else branch keeps the context stashed and does the deletion.

Reply:
Yeah, I know but I wasn't able to comment there as that wasn't a modified code section. In summary, I was just wondering if we should add context restore in multiple places? Or should we just stash again before performing any DB call. That will keep the stash context scope smaller and more secured IMO.

Reply:
The previous code used `runBefore()` on the action listener which basically just delayed restoring it until the action lister was completed (with success or failure).   I was somewhat trying to restore earlier with the initial code but it was too early.  This delays it for when it's needed.

I can probably put it back closer to the `runBefore()` implementation which is the opposite direction of your wondering... or since it's used in try-with-resources, it actually will be automatically restored as part of that auto-closure, so manually doing anything to it isn't actually required here.

I'm happy to go either way, which do you prefer?

Reply:
I agree with you. Let's put it back closer to the `runBefore()` and then we can make the changes in main to keep the context scope smaller.